third_party/proton/csrc/lib/Profiler/CuptiProfiler.cpp

```diff
@@ -199,17 +237,66 @@ void CuptiProfiler::CuptiProfilerPimpl::callbackFn(void *userData,
                                                    CUpti_CallbackDomain domain,
                                                    CUpti_CallbackId cbId,
                                                    const void *cbData) {
-  CuptiProfiler &profiler = profilerState.profiler;
-  const CUpti_CallbackData *callbackData =
-      reinterpret_cast<const CUpti_CallbackData *>(cbData);
-  if (callbackData->callbackSite == CUPTI_API_ENTER) {
-    auto scopeId = Scope::getNewScopeId();
-    profilerState.record(scopeId);
-    profilerState.enterOp(scopeId);
-    profiler.correlation.correlate(callbackData->correlationId);
-  } else if (callbackData->callbackSite == CUPTI_API_EXIT) {
-    profilerState.exitOp();
-    profiler.correlation.submit(callbackData->correlationId);
+  CuptiProfiler &profiler = threadState.profiler;
+  if (domain == CUPTI_CB_DOMAIN_RESOURCE) {
+    auto *resourceData =
+        reinterpret_cast<CUpti_ResourceData *>(const_cast<void *>(cbData));
+    auto *graphData =
+        reinterpret_cast<CUpti_GraphData *>(resourceData->resourceDescriptor);
+    auto *pImpl = dynamic_cast<CuptiProfilerPimpl *>(profiler.pImpl.get());
+    uint32_t graphId = 0;
+    uint32_t graphExecId = 0;
+    if (graphData->graph)
```

Comment:
Is it worth doing any conditional-based checking here to make sure the casts above were successful? Are there causes in which they could fail?